src/Middleware/RequestLimiter/src/RequestLimiterMiddleware.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.ResourceLimits;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.RequestLimiter
+{
+    public class RequestLimiterMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ILogger _logger;
+        private readonly RequestLimiterOptions _options;
+
+        public RequestLimiterMiddleware(RequestDelegate next, ILoggerFactory loggerFactory, IOptions<RequestLimiterOptions> options)
+        {
+            _next = next;
+            _logger = loggerFactory.CreateLogger<RequestLimiterMiddleware>();
+            _options = options.Value;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            _logger.LogInformation("Resource limiting: " + context.Request.Path);
+
+            var endpoint = context.GetEndpoint();
+            var attributes = endpoint?.Metadata.GetOrderedMetadata<RequestLimitAttribute>();
+
+            if (attributes == null)
+            {
+                await _next.Invoke(context);
+                return;
+            }
+
+            var resourceLeases = new Stack<ResourceLease>();
+            try
+            {
+                foreach (var attribute in attributes)
+                {
+                    if (!string.IsNullOrEmpty(attribute.Policy) && attribute.Limiter != null)
+                    {
+                        throw new InvalidOperationException("Cannot specify both policy and limiter");
+                    }
```

Comment:
Why not allow both? The logic below seems like it would work as-is.

What if both are supplied on separate attributes?

These per-request exceptions aren't great, is there any way to validate this data on startup?

Reply:
Theoretically, this can never happen. When the Attribute is constructed, only one of Policy or Limiter is set. I don't think it makes sense to use both on an Attribute. If both are required, two separate attributes should added, would that be sufficient?

Reply:
Then why even check for the condition?

Reply:
Good point, it felt weird to not explicitly check this condition. Maybe a Debug.Assert instead is better.